common/quotas/multi_reservation_impl.go

```diff
@@ -0,0 +1,96 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package quotas
+
+import (
+	"time"
+)
+
+type (
+	MultiReservationImpl struct {
+		ok           bool
+		reservations []Reservation
+	}
+)
+
+var _ Reservation = (*MultiReservationImpl)(nil)
+
+func NewMultiReservation(
+	ok bool,
+	reservations []Reservation,
+) *MultiReservationImpl {
+	return &MultiReservationImpl{
+		ok:           ok,
+		reservations: reservations,
+	}
+}
+
+// OK returns whether the limiter can provide the requested number of tokens
+func (r *MultiReservationImpl) OK() bool {
+	return r.ok
+}
+
+// Cancel indicates that the reservation holder will not perform the reserved action
+// and reverses the effects of this Reservation on the rate limit as much as possible
+func (r *MultiReservationImpl) Cancel() {
+	r.CancelAt(time.Now())
+}
+
+// CancelAt indicates that the reservation holder will not perform the reserved action
+// and reverses the effects of this Reservation on the rate limit as much as possible
+func (r *MultiReservationImpl) CancelAt(now time.Time) {
+	if !r.ok {
```

Comment:
[Nit] if somebody has already called Cancel() once, doing a second Cancel() call will have no effect, right? Or should we set a canceledFlag and bail/out (and or set ok to false as a part of this CancelAt call)

Reply:
the reservation implementations in the quotas package do not expect the caller to call cancel more than 1 times
similar behavior also applies to go's impl (which shares the same interface):
https://github.com/golang/time/blob/f8bda1e9f3ba/rate/rate.go#L154